### PR TITLE
Roy/bam c

### PR DESCRIPTION
**Pull request by royster57**

### CODE REVIEW

File: common/src/chain/config.rs [hunk 1]
```diff
@@ -117,6 +117,21 @@ pub fn create_mainnet() -> ChainConfig {
     }
 }
 
+pub fn create_mainnet_with_netupgrades(net_upgrades: NetUpgrades<UpgradeVersion>) -> ChainConfig {
```
Comment (TheQuantumPhysicist) [inline common/src/chain/config.rs:120]: This is very dangerous. Mainnet parameters should not be parametrized in any way and should be 100% determined at compile-time. If this is for testing, please create a separate net for unit tests like we did today.

Comment (royster57) [reply common/src/chain/config.rs]: Removed it.

File: consensus/src/detail/consensus_validator.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    fn get_ancestor(
+        &self,
+        block_index: &BlockIndex,
+        ancestor_height: BlockHeight,
+    ) -> Result<BlockIndex, BlockError>;
+}
+
+pub(crate) fn validate_consensus(
+    chain_config: &ChainConfig,
+    block: &Block,
```
Comment (TheQuantumPhysicist) [inline consensus/src/detail/consensus_validator.rs:46]: The header should be enough, not the whole block.

File: consensus/src/detail/consensus_validator.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+fn do_validate(
+    chain_config: &ChainConfig,
+    block: &Block,
+    block_consensus_data: &ConsensusData,
+    consensus_status: &ConsensusStatus,
+    block_index_handle: &dyn BlockIndexHandle,
+) -> Result<(), BlockError> {
+    match (block_consensus_data, consensus_status) {
```
Comment (TheQuantumPhysicist) [inline consensus/src/detail/consensus_validator.rs:80]: Nope. Please don't match multiple things together. Please match one, then pass the result to another function that does the mapping to an error or a valid result.

Also, requiring no consensus and providing consensus is invalid.

File: consensus/src/detail/consensus_validator.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        block_index: &BlockIndex,
+        ancestor_height: BlockHeight,
+    ) -> Result<BlockIndex, BlockError>;
+}
+
+pub(crate) fn validate_consensus(
+    chain_config: &ChainConfig,
+    block: &Block,
+    block_index_handle: &dyn BlockIndexHandle,
+) -> Result<(), BlockError> {
```
Comment (TheQuantumPhysicist) [inline consensus/src/detail/consensus_validator.rs:48]: We also need an accessor for the utxo set. This is not a priority and we can postpone this whole thing. But feel free to do it if you have nothing else to do.

The accessor should have the functions:

- Check if a utxo is spent
- Check if a utxo is spendable

Currently this information can be retrieved from the transaction index.

File: common/src/chain/upgrades/netupgrade.rs [hunk 5]
```diff
@@ -25,17 +35,55 @@ pub trait Activate {
...
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub enum ConsensusUpgrade {
+    PoW { initial_difficulty: Compact },
+    PoS,
+    DSA,
+    IgnoreConsensus,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub enum ConsensusStatus {
```
Comment (TheQuantumPhysicist) [inline common/src/chain/upgrades/netupgrade.rs:51]: I would call this `RequiredConsensus`

File: consensus/src/detail/error.rs [hunk 6]
```diff
@@ -81,7 +90,15 @@ pub enum BlockError {
...
     #[error("Unexpected numeric type conversion error `{0:?}`")]
     InternalNumTypeConversionError(Id<Block>),
-    // To be expanded
+    #[error("Conversion failed: `{0:?}`")]
+    Conversion(String),
+    #[error("No PoW data for block")]
+    NoPowData,
+    #[error("Block consensus data is {block_consensus_data:?}, current config is {chain_consensus_status:?}")]
+    ConsensusTypeMismatch {
+        block_consensus_data: ConsensusData,
```
Comment (TheQuantumPhysicist) [inline consensus/src/detail/error.rs:99]: Consensus data is relatively large. You can't print it in an error. You can make a string value out of it that mentions the type.

File: consensus/src/detail/consensus_validator.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        &consensus_status,
+        block_index_handle,
+    )?;
+    Ok(())
+}
+
+fn do_validate(
+    chain_config: &ChainConfig,
+    block: &Block,
+    block_consensus_data: &ConsensusData,
```
Comment (TheQuantumPhysicist) [inline consensus/src/detail/consensus_validator.rs:76]: Isn't it redundant to pass both the block and its consensus data? You can already extract the consensus data from the block itself.

File: common/src/chain/config.rs [hunk 8]
```diff
@@ -117,6 +116,21 @@ pub fn create_mainnet() -> ChainConfig {
     }
 }
 
+pub fn unit_test_config() -> ChainConfig {
```
Comment (TheQuantumPhysicist) [inline common/src/chain/config.rs:119]: `create_unit_test_config()`. Let's follow the same convention.

File: consensus/src/detail/mod.rs [hunk 9]
```diff
@@ -567,9 +608,7 @@ impl<'a> ConsensusRef<'a> {
     }
 
     fn check_consensus(&self, block: &Block) -> Result<(), BlockError> {
```
Comment (TheQuantumPhysicist) [inline consensus/src/detail/mod.rs:610]: This function can be eliminated. No need for one liners in this context.

**Comment by TheQuantumPhysicist** [comment]
Overall this is starting to look like what I envisioned. I like the isolation you created. Please continue and see the comments I made.

File: consensus/src/detail/consensus_validator.rs (hunk 2 above)
Comment (royster57) [reply consensus/src/detail/consensus_validator.rs:26]: Done.

File: consensus/src/detail/consensus_validator.rs (hunk 3 above)
Comment (royster57) [reply consensus/src/detail/consensus_validator.rs:53]: The IgnoreConsensus variant is so tests don't trip up for not having consensus data... That's why we return Ok when we encounter this variatn.

File: common/src/chain/upgrades/netupgrade.rs (hunk 5 above)
Comment (royster57) [reply common/src/chain/upgrades/netupgrade.rs]: Done.

File: consensus/src/detail/error.rs (hunk 6 above)
Comment (royster57) [reply consensus/src/detail/error.rs]: Done.

File: consensus/src/detail/consensus_validator.rs (hunk 3 above)
Comment (TheQuantumPhysicist) [reply consensus/src/detail/consensus_validator.rs]: I think there's a misunderstanding here. I meant, IF netupgrades says no consensus is required, then adding consensus is a wrong.

File: common/src/chain/config.rs (hunk 8 above)
Comment (royster57) [reply common/src/chain/config.rs:119]: Done.

File: consensus/src/detail/mod.rs (hunk 9 above)
Comment (royster57) [reply consensus/src/detail/mod.rs:610]: Done.

File: consensus/src/detail/consensus_validator.rs (hunk 7 above)
Comment (royster57) [reply consensus/src/detail/consensus_validator.rs]: Changed this around.

File: consensus/src/detail/consensus_validator.rs (hunk 3 above)
Comment (royster57) [reply consensus/src/detail/consensus_validator.rs]: Gotcha.

File: common/src/primitives/compact.rs [hunk 10]
```diff
@@ -129,4 +129,19 @@ mod tests {
         err_conversion(0x01fedcba);
         err_conversion(!0x00800000); // overflow
     }
+    /*
```
Comment (TheQuantumPhysicist) [inline common/src/primitives/compact.rs:132]: ?

File: consensus/src/detail/consensus_validator.rs (hunk 4 above)
Comment (TheQuantumPhysicist) [reply consensus/src/detail/consensus_validator.rs:30]: I'm resolving this as it's gonna be done for PoS at some point later.

**Review by TheQuantumPhysicist — APPROVED** [review]
Looking good

File: consensus/src/detail/pow/work.rs [hunk 11]
```diff
@@ -17,35 +17,83 @@
...
+    pow_status: &PoWStatus,
+    block_index_handle: &dyn BlockIndexHandle,
+) -> Result<Compact, BlockError> {
+    match pow_status {
+        PoWStatus::Threshold { initial_difficulty } => Ok(*initial_difficulty),
+        PoWStatus::Ongoing => {
+            let prev_block_id = header
+                .get_prev_block_id()
+                .clone()
+                .expect("If PoWStatus is `Onging` then we cannot be at genesis");
```
Comment (altonen) [inline consensus/src/detail/pow/work.rs:81]: Onging

**Review by altonen — APPROVED** [review]

**Review by muursh — APPROVED** [review]

**Pull request merged**